Addressing most of the comments about src/dagmc files

bam241 opened the pull request:
This addresses most of the comments about dagmc sources from #660 .

I was not sure how to fix some of them so I ping @pshriwise @gonuke @makeclean on them to get more information.

I also included changes suggested by my linter (mostly line <80 chars...)

bam241 commented:
I am thinking that addressing the linter suggestion in this PR might be a bit much...
Let me know I can simplify this if necessary....

bam241 committed:
addressing most of the PullRequest-Agent comments in src/dagmc folder

astyling

astyling dagMC.hpp

this should do it

this should finally be working

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -92,16 +94,22 @@ float DagMC::version(std::string* version_string) {
-  const char* interface_string = DAGMC_INTERFACE_REVISION;
-  if (strlen(interface_string) >= 5) {
+  std::string interface_string = DAGMC_INTERFACE_REVISION;
+  if (interface_string.rfind("$Rev:") != std::string::npos) {
```
Does anyone know if this actually works!?!?  This looks like a tag expansion that was part of CVS/SVN, but I don't think we ever get "$Rev:" in the file anymore.  Perhaps we need a new solution for this entirely?

gonuke replied:
If it is working, maybe we should define `$Rev:` as a string variable, and refer to it's length where we otherwise use 5 as a magic number below.

I'm not sure why 10 as a magic number below?

bam241 replied:
sounds about right for `$rev`

I don't know what this 10 is for....

bam241 replied:
this comes from `src/dagmc/DagMCVersion.hpp.in`

bam241 replied:
and did not get populate in my `DagMCVersion.hpp`:
`#define DAGMC_INTERFACE_REVISION "$Rev$"` <- I still have this like in `bld/src/dagmc/DagMCVersion.hpp`

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -92,16 +94,22 @@ float DagMC::version(std::string* version_string) {
+    return result;
   }
-  return result;
+  /* SECTION I: Geometry Initialization and problem setup */
```
Why did this comment get move into this function?

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -111,7 +119,12 @@ ErrorCode DagMC::load_file(const char* cfile) {
 
   // get the last 4 chars of file .i.e .h5m .sat etc
-  memcpy(file_ext, &cfile[strlen(cfile) - 4], 4);
+  if (sizeof(cfile) / sizeof(char) > 4) {
```
add variable `int file_extension_length = 4` instead of magic number

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -111,7 +119,12 @@ ErrorCode DagMC::load_file(const char* cfile) {
+  if (sizeof(cfile) / sizeof(char) > 4) {
+    memcpy(file_ext, &cfile[strlen(cfile) - 4], 4);
+  } else {
+    std::cerr << "DagMC warning: unhandled file "
```
It might still be better for this error to be generated later, after any error messages that might get generated by `load_file`.  It also looks like they create a `std::string` version of `cfile` - maybe we should do that earlier and take advantage of `std::string` functions throughout.

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -379,10 +391,21 @@ ErrorCode DagMC::build_indices(Range& surfs, Range& vols) {
-  EntityHandle tmp_offset = std::max(surfs.back(), vols.back());
+  EntityHandle tmp_offset = 0;
+
+  if (surfs.size() != 0 && vols.size() != 0) {
```
Looking back at the Pull-Request comment....  I am not sure if these checks are necessary.  All valid DagMC geometries must have at least one surface and at least one volume.  I don't know if there is a prior check to make sure this is true.  If there is, then no check is needed here.  If there is not, then perhaps we should check that both are non-zero length and give a failing error if either one is, and then use the (nearly) original code....

bam241 replied:
just wondering if only one of them is non zero would the `min()`/`max()` still work ?

gonuke commented on src/dagmc/DagMC.hpp:
```diff
@@ -359,9 +364,12 @@ class DagMC {
-  std::vector<EntityHandle>& vol_handles() {return entHandles[3];}
-  std::vector<EntityHandle>& group_handles() {return entHandles[4];}
+  std::vector<EntityHandle>& surf_handles()
+  {return entHandles[surfs_handle_idx];}
```
Not sure I like this style...  is this what Astyle requires here? I'd prefer some kind of indent unless our style guide disallows it???

bam241 replied:
`astyle` wouldn't let me indent it.
So I opened the brackets.

ljacobson64 replied:
I just ran astyle on

```
std::vector<EntityHandle>& surf_handles() {return entHandles[2];}
std::vector<EntityHandle>& vol_handles() {return entHandles[3];}
std::vector<EntityHandle>& group_handles() {return entHandles[4];}
```

 and it gave me

```
std::vector<EntityHandle>& surf_handles() {
    return entHandles[2];
}
std::vector<EntityHandle>& vol_handles() {
    return entHandles[3];
}
std::vector<EntityHandle>& group_handles() {
    return entHandles[4];
}
```

ljacobson64 replied:
Not to turn this into a discussion about changing our preferred code formatter, but `clang-format` gives

```
std::vector<EntityHandle> &surf_handles() { return entHandles[2]; }
std::vector<EntityHandle> &vol_handles() { return entHandles[3]; }
std::vector<EntityHandle> &group_handles() { return entHandles[4]; }
```

which looks the best in my opinion

bam241 replied:
Yep this is what I ended up implementing :)

as it is in the header file, I was trying to keep it as short as possible.... (this is why I tried the open/close brackets in the same line.

bam241 replied:
As I replaced the magic number (`2`/`3`/`4`) by variable, lines are now `> 80` chars with the inline formatting

gonuke commented on src/dagmc/DagMC.hpp:
```diff
@@ -446,7 +454,7 @@ inline int DagMC::index_by_handle(EntityHandle handle) {
 
 inline unsigned int DagMC::num_entities(int dimension) {
-  assert(0 <= dimension && 3 >= dimension);
+  assert(0 <= dimension && groups_handle_idx >= dimension);
```
Since you've setup these variables, you might want to add:
`vertex_handle_idx = 0` (to use here) and
`curve_handle_idx = 1`  (for completeness)

gonuke commented on src/dagmc/dagmcmetadata.cpp:
```diff
@@ -135,81 +138,82 @@ void dagmcMetaData::parse_material_data() {
-    } else {
-      grp_name = "mat:" + material_props[0];
-      volume_density_data_eh[eh] = "";
+    if (density_props.size() == 1) {
```
I wonder if it's possible for this to be valid if >1???

bam241 replied:
I do not know.

Would anyone would have with some more experience about this ?

makeclean replied:
> 1 is caught earlier on, line 102

bam241 replied:
ok I remember now,
162 ensure that `density_props .size() <= 1`
this ensure we have at lease 1 element before calling `density_props[0]`

gonuke commented on src/dagmc/dagmcmetadata.cpp:
```diff
@@ -250,7 +254,15 @@ void dagmcMetaData::parse_importance_data() {
+        errno = 0;
+        double imp = std::strtod(pair.second.c_str(), &e);
+        if (*e != '\0' || errno != 0) {
+          std::cerr << "Can't parse " << pair.second << " as a float."
```
Provide the user a little more insight about where this came up.
```suggestion
          std::cerr << "Can't parse importance " << pair.second << " as a float."
```

gonuke commented on src/dagmc/dagmcmetadata.cpp:
```diff
@@ -250,7 +254,15 @@ void dagmcMetaData::parse_importance_data() {
       // insert into map too
       if (importance_map[eh].count(pair.first) == 0) {
-        importance_map[eh][pair.first] = atof(pair.second.c_str());
+        char* e;
```
Wow - this is verbose.  Why can't C++ make this easier???  Does `std::stod` do any better?

makeclean replied:
It probably can, IIRC we hadnt shifted off the C++03 standard for MOAB compatability, so going back to C99 was the only option

gonuke commented on src/dagmc/tests/dagmc_pointinvol_test.cpp:
```diff
@@ -57,7 +57,7 @@ int dagmc_point_in_vol_dir(double origin[3], double dir[3], int vol_idx) {
   EntityHandle next_surf;
 
-  // normalise the vector
+  // normalise the direction vector
```
```suggestion
  // direction vectors are always interpreted as unit vectors - make sure this one is normalized
```

gonuke reviewed (changes requested):
Some thoughts based on context and background.

gonuke commented:
Not sure why this is failing??  Might need a little more verbosity to figure it out.

bam241 committed:
Apply suggestions from code review

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
adressing most of @gonuke comments

adding comments on error checking the string atof

changing strstod to stod and adding mechanisum to catch conversion exception

astyling DagMC.cpp and dagmcmetadata.cpp

astyling DagMC.hpp

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -92,27 +94,43 @@ float DagMC::version(std::string* version_string) {
   return result;
 }
 
-/* SECTION I: Geometry Initialization and problem setup */
```
Did we take out the other comments that for these `Section` headers?

bam241 replied:
I am not sure what I did with this...
Some kind of a weird combination of typos...

I'll restore it.

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -379,10 +396,14 @@ ErrorCode DagMC::build_indices(Range& surfs, Range& vols) {
-  // max
-  EntityHandle tmp_offset = std::max(surfs.back(), vols.back());
+  if (surfs.size() == 0 || vols.size() == 0) {
+    std::cout << "Volumes or Surfaces founds" << std::endl;
```
``` suggestion
    std::cout << "Volumes or Surfaces not found" << std::endl;
```

gonuke reviewed (changes requested):
Maybe a couple of things lost in the revisions...

gonuke reviewed (changes requested):
Maybe a couple of things lost in the revisions...

bam241 committed:
Apply suggestions from code review

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bam241 committed:
Adressing remaining review comments

addressing @gonuke last comments

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -124,7 +144,7 @@ ErrorCode DagMC::load_file(const char* cfile) {
     // Print a warning if an option was unhandled for a file that does not end in '.h5m'
     std::string filename(cfile);
-    if (filename.length() < 4 || filename.substr(filename.length() - 4) != ".h5m") {
+    if (std::string(file_ext) != ".h5m") {
```
`file_ext` is already a `std::string` - no need to caste it here

gonuke commented on src/dagmc/DagMC.cpp:
```diff
@@ -379,10 +398,14 @@ ErrorCode DagMC::build_indices(Range& surfs, Range& vols) {
 
   // surf/vol offsets are just first handles
-  setOffset = std::min(*surfs.begin(), *vols.begin());
+  EntityHandle tmp_offset = 0;
```
Combine this with line 408

gonuke reviewed (changes requested):
Two small requests & rebase for merge today

bam241 commented:
@gonuke I think this is ready.
I rebased, addressed your 2 remaining comments and cleaned the commit history (squash).

gonuke commented:
Thanks @bam241 - finally merging!!!
